ops/jujucontext.py

```diff
@@ -0,0 +1,345 @@
+
+    For example 'cdac5656-2423-4388-8f30-41854b4cca7d' (from JUJU_MODEL_UUID).
+    """
+
+    notice_id: Optional[str] = None
+    """The ID of the notice.
+
+    For example '1', (from JUJU_NOTICE_ID).
+    """
+
+    notice_key: Optional[str] = None
+    """The key of the notice.
+
+    For example 'example.com/a', (from JUJU_NOTICE_KEY).
+    """
+
+    notice_type: Optional[str] = None
+    """The type of the notice.
+
+    For example 'custom' (from JUJU_NOTICE_TYPE).
+    """
+
+    pebble_check_name: Optional[str] = None
+    """The name of the pebble check.
+
+    For example 'http-check' (from JUJU_PEBBLE_CHECK_NAME).
+    """
+
+    relation_departing_unit_name: Optional[str] = None
+    """The unit that is departing a relation.
+
+    For example 'remote/42' (from JUJU_DEPARTING_UNIT).
+    """
+
+    relation_name: Optional[str] = None
+    """The name of the relation.
+
+    For example 'database' (from JUJU_RELATION).
+    """
+
+    relation_id: Optional[int] = None
+    """The id of the relation.
+
+    For example 1 (integer) if the original environment variable's value is 'database:1'
+    (from JUJU_RELATION_ID).
+    """
+
+    remote_app_name: Optional[str] = None
+    """The name of the remote app.
+
+    For example 'remoteapp1' (from JUJU_REMOTE_APP).
+    """
+
+    remote_unit_name: Optional[str] = None
+    """The name of the remote unit.
+
+    For example 'remoteapp1/0' (from JUJU_REMOTE_UNIT).
+    """
+
+    secret_id: Optional[str] = None
+    """The ID of the secret.
+
+    For example 'secret:dcc7aa9c-7202-4da6-8d5f-0fbbaa4e1a41' (from JUJU_SECRET_ID).
+    """
+
+    secret_label: Optional[str] = None
+    """The label of the secret.
+
+    For example 'db-password' (from JUJU_SECRET_LABEL).
+    """
+
+    secret_revision: Optional[int] = None
+    """The revision of the secret.
+
+    For example 42 (integer) (from JUJU_SECRET_REVISION).
+    """
+
+    storage_name: Optional[str] = None
+    """The storage name.
+
+    For example 'my-storage' if the original environment variable's value is 'my-storage/1'
+    (from JUJU_STORAGE_ID).
+    """
+
+    unit_name: str = ''
+    """The name of the unit.
+
+    For example 'myapp/0' (from JUJU_UNIT_NAME).
+    """
+
+    version: JujuVersion = dataclasses.field(default_factory=lambda: JujuVersion('0.0.0'))
+    """The version of Juju.
+
+    For example '3.4.0' (from JUJU_VERSION).
+    """
+
+    workload_name: Optional[str] = None
+    """The name of the workload.
+
+    For example 'workload' (from JUJU_WORKLOAD_NAME).
+    """
+
+    @classmethod
+    def from_dict(cls, env: Mapping[str, Any]) -> '_JujuContext':
+        return _JujuContext(
+            action_name=env.get('JUJU_ACTION_NAME') or None,
+            action_uuid=env.get('JUJU_ACTION_UUID') or None,
+            charm_dir=(
+                Path(f'{__file__}/../../..').resolve()
+                if not env.get('JUJU_CHARM_DIR')
+                else Path(env['JUJU_CHARM_DIR']).resolve()
```

Comment:
Let's avoid the not/negative here (simpler, and to match the other similar constructs):

```suggestion
                Path(env['JUJU_CHARM_DIR']).resolve()
                if env.get('JUJU_CHARM_DIR')
                else Path(f'{__file__}/../../..').resolve()
```